jobserver/commands/project_members.py

```diff
@@ -0,0 +1,61 @@
+from django.db import transaction
+
+from ..authorization.utils import dotted_path
+from ..models import AuditableEvent
+
+
+@transaction.atomic()
+def add(*, project, users, roles, by):
+    for user in users:
+        membership = project.memberships.create(
+            user=user,
+            created_by=by,
+            roles=roles,
+        )
+
+        # create this here because we can't pass a reference to the membership
+        # before it's created
+        AuditableEvent.objects.create(
+            type=AuditableEvent.Type.PROJECT_MEMBER_ADDED,
+            target_model=membership._meta.label,
+            target_id=membership.pk,
+            changed_by=by.username,
+            new=user.username,
+        )
+
+
+@transaction.atomic()
+def update_roles(*, member, by, roles):
+    AuditableEvent.objects.create(
+        type=AuditableEvent.Type.PROJECT_MEMBER_UPDATED_ROLES,
+        old=",".join([dotted_path(r) for r in member.roles]),
+        target_model=member._meta.label,
+        target_field="roles",
+        target_id=member.pk,
+        changed_by=by.username,
+        new=",".join([dotted_path(r) for r in roles]),
+    )
+
+    member.roles = roles
+    member.save(update_fields=["roles"])
+
+
+@transaction.atomic()
+def remove(*, membership, by):
+    membership.delete()
```

Comment:
Deleting the membership appears to remove all previous audit events for that user, but only on the project audit log.

STR:
1. Add Millie to a project -> one project log entry
2. Edit her role -> two project log entries
3. Remove Millie from the project -> one project log entry showing that the user was removed.

My own user log is closer to being correct, although it's missing Mille's name in one case. Millie's user log doesn't show the change in roles.

![Screenshot from 2023-10-06 18-09-28](https://github.com/opensafely-core/job-server/assets/3889554/3803e033-660c-473e-90e8-454d342b380e)

![Screenshot from 2023-10-06 18-08-20](https://github.com/opensafely-core/job-server/assets/3889554/7597a750-dd95-420f-9350-fde9f9d9fb03)

![Screenshot from 2023-10-06 18-11-57](https://github.com/opensafely-core/job-server/assets/3889554/4b3a26ef-d8f1-4503-9dd8-121e4fb60724)

Reply:
Well this was a pain to fix, deleting a member caused all sorts of issues.

I've reworked the data model to store a parent model&ID, with the idea that we'll nearly always store Project here but it's left open to track anything.  I've also added a `target_user` field for usernames an event happens to, rather than using new/old since if we have that information it's somewhat of a special case.  These changes ended up massively simplifying the presenters which feels like a good signal we're moving in the right direction.

Roles added with a membership are now tracked as a separate event record to the membership addition so as not to make the model too confusing as we add more event types, ie if we add roles now, what are we going to need in the future?

This covers doing something like:
* add member with N roles
* edit members roles
* delete member